lib/ElementHandle.js

```diff
@@ -137,24 +151,72 @@ class ElementHandle extends JSHandle {
     return {x, y, width, height};
   }
 
+  /**
+   * @return {!Promise<?{x: number, y: number, width: number, height: number}>}
+   */
+  async _assertBoundingBox() {
+    const boundingBox = await this.boundingBox();
+    if (!boundingBox)
+      throw new Error('Node is not visible');
+    return boundingBox;
+  }
+
   /**
    *
    * @param {!Object=} options
    * @returns {!Promise<Object>}
    */
   async screenshot(options = {}) {
-    await this._scrollIntoViewIfNeeded();
+    let needsViewportReset = false;
+
+    let boundingBox = await this._assertBoundingBox();
+
+    const viewport = Object.assign({}, this._page.viewport());
+
+    const newViewport = {
+      width: Math.max(viewport.width, Math.ceil(boundingBox.width)),
+      height: Math.max(viewport.height, Math.ceil(boundingBox.height)),
+    };
+
+    const initialScroll = await this.executionContext().evaluate(function() {
+      return { scrollX: window.scrollX, scrollY: window.scrollY };
+    });
+
+    if (newViewport.width !== viewport.width || newViewport.height !== viewport.height) {
+      await this._page._emulationManager.emulateViewport(this._client, newViewport);
+
+      needsViewportReset = true;
+    }
+
+    await this.executionContext().evaluate(function(el) {
+      const box = el.getBoundingClientRect();
+
+      window.scrollTo(
+          box.left + window.scrollX,
+          box.top + window.scrollY
+      );
+    }, this);
+
+    boundingBox = await this._assertBoundingBox();
+
     const { layoutViewport: { pageX, pageY } } = await this._client.send('Page.getLayoutMetrics');
 
-    const boundingBox = await this.boundingBox();
-    if (!boundingBox)
-      throw new Error('Node is not visible');
     const clip = Object.assign({}, boundingBox);
     clip.x += pageX;
     clip.y += pageY;
-    return await this._page.screenshot(Object.assign({}, {
+
+    const imageData = await this._page.screenshot(Object.assign({}, {
       clip
     }, options));
+
+    if (needsViewportReset)
+      await this._page._emulationManager.emulateViewport(this._client, viewport);
+
+    await this.executionContext().evaluate(function(x, y) {
```

Comment:
I wouldn't worry about putting the scroll position back. I'd favor minimal interactions with the page over cleaning up after ourselves.

Reply:
Fixed